pubtools/pulplib/_impl/model/repository/file.py

```diff
@@ -20,3 +23,53 @@ class FileRepository(Repository):
     )
 
     mutable_urls = attr.ib(default=attr.Factory(lambda: ["PULP_MANIFEST"]), type=list)
+
+    def upload_file(self, filename, relative_url=None):
+        """Upload a file to this repository.
+
+        The upload operation includes 4 steps:
+            1. Request an upload id from pulp
+            2. With the requested upload id, upload the file
+            3. After the upload's done, import the uploaded file to repository
+            4. Delete the upload request
+
+        Args:
+            filename (str)
+                Path to a local file to upload.
+
+            relative_url (str)
+                Path that should be used in remote repository
+
+                if omitted, filename will be used.
+        Returns:
+            Future[:class:`~pubtools.pulplib.Task`]
+                A future which is resolved when publish succeeds.
+
+                The future contains the task to delete upload request
+
+        Raises:
+            DetachedException
+                If this instance is not attached to a Pulp client.
+        """
+        if not self._client:
+            raise DetachedException()
+
+        path, name = os.path.split(filename.rstrip("/"))
+        if not relative_url:
+            relative_url = path
+
+        # request upload id and wait for it
+        upload_id = self._client._request_upload().result()["upload_id"]
+
+        # caculate hash/size of the file and upload it to pulp
+        upload_fs, checksum, size = self._client._do_upload_file(
```

Comment:
It's an odd mix of blocking and non-blocking here.  What was the point of implementing the private method _do_upload_file as non-blocking if the only place you use it in public API is to immediately block on the result anyway?

Reply:
I wrote it to blocking since import can't be started until the file's uploaded to pulp, am I right?

Reply:
Yeah, import can't be started until after the file's uploaded, but that doesn't mean you should use result() to block.  If you do that then it means the caller can't upload more than one file concurrently.

Here's a simplified pseudocode example.  In non-blocking style:

```
def upload_file(content, filename):
  upload_id_f = client._request_upload()
  # Once we have upload id then call upload_content
  upload_complete_f = f_flat_map(upload_id_f, lambda upload_id: client._upload_content(upload_id, content))
  # Once upload is complete then call import_content
  import_complete_f = f_flat_map(upload_complete_f, lambda _: client._import_content(upload_id_f.result(), ...))
  # Once import is complete then call delete_upload (but we don't care about the result)
  f_map(import_complete_f, lambda _: client._delete_upload(upload_id_f.result()))

  # Where are we now:
  # - we haven't necessarily *done* anything yet, just queued up the whole sequence
  # - hence the caller can immediately call us again to queue any more needed uploads as needed
  # - we'll return a future which will return whatever _import_content returned
  return import_complete_f
```

Reply:
ah, I finally got what you mean. I didn't expect we could wrap things in such a way, it's really nice.